### ISSUE TITLE
Add bgdx

### ISSUE THREAD

**Issue by fire** [ruslo/hunter#372]
I'm writing a cmake'fied version of bgfx. What do I need to do to get it into hunter?

https://github.com/fire/bgfx

**Comment by fire** [comment]
Requires the zip to have submodules which aren't included by default and causes the fetch to break.

**Comment by ruslo** [comment]
I see a lot of 3rd party stuff so it would be nice to have separate package for each of them:
* https://github.com/bkaradzic/bx/tree/3db8287bbb38a31005bc7fe5c38d0edfe5402c6d/3rdparty
* https://github.com/bkaradzic/bgfx/tree/2288f2e779538e8dfeb3880d5d4c1f4275bd56ed/3rdparty

There is GitHub organization for forked projects:
* https://github.com/hunter-packages

Feel free to send me a request for the needed forks, I will create them and give you permission for modifications.

**Comment by fire** [comment]
As I am not the author of the software, I'm not sure what I can do to separate the third party packages. Would git subtrees be a method?

BX and BGFX depend on each other and the author prefers the GENIE build system. So I don't think the author will support cmake hunter.

**Comment by ruslo** [comment]
> As I am not the author of the software

I guess having different roles for maintainer of the package and author of the software is a quite usual situation even for other package managers, not only for Hunter.

> Would git subtrees be a method?

I'm not sure I understand how it can be helpful. As a final result we need some downloadable archive which we can used to build a package. Having `3rdparty` directory in repo usually means a workaround for not having standardized package manager. It may works at a start but we may have conflicts in future (imaging two packages `Foo` and `Boo` which have different version of package `A` in `3rdparty` directory). That's why I say that it **would** be nice to create separate package for them.

> BX and BGFX depend on each other and the author prefers the GENIE build system. So I don't think the author will support cmake hunter

If you don't want to use CMake as a build tool it's possible to create "custom build scheme". But from my experience it most of the cases supporting CMake-ified package is easier.

**Comment by fire** [comment]
Abandoning this. I will be working on urho3d integration.

**Comment by ruslo** [comment]
> I will be working on urho3d integration.

I've openned https://github.com/ruslo/hunter/issues/389

**Comment by headupinclouds** [comment]
I'm also interested in a lightweight cross platform game/rendering engine, primarily for iOS and Android (w/ OS X/Linux/Windows for debugging and development).  I was planning to look at gameplay3d (#188) but would help with an urho3d hunterization effort if that has some momentum.  I have almost no familiarity with (or preference for) either package at this point.

**Comment by nick70** [comment]
[bgfx.cmake](https://github.com/JoshuaBrookover/bgfx.cmake) provides a cmake wrapper for bgfx.

It still uses submodules and there are no zip releases but its a step closer to what we need.

**Comment by spacepluk** [comment]
I might be wrong, but I believe the stuff under 3rdparty has patches and bgfx it's not guaranteed to work with the upstream versions. (cc @bkaradzic)

**Comment by OvermindDL1** [comment]
> I might be wrong, but I believe the stuff under 3rdparty has patches and bgfx is not guaranteed to work with the upstream versions.

That sounds buggy... why not commit upstream?  And why is genie *still* being used by bgfx...  >.>

**Comment by ruslo** [comment]
> I might be wrong, but I believe the stuff under 3rdparty has patches and bgfx is not guaranteed to work with the upstream versions

You can use custom `config.cmake` for the packages that is not compabitle with upstream, e.g. see Urho3D: https://docs.hunter.sh/en/latest/packages/pkg/Urho3D.html

**Comment by spacepluk** [comment]
@ruslo can I make the bgfx package default to these custom versions? and where do I define these versions? I'm gonna try to package this.

**Comment by spacepluk** [comment]
There's another thing that I'm not sure how to manage.  Since bgfx doesn't have tagged releases I was thinking of using the git SHAs of the packaged versions in the hunter_add_version calls.  Is there a better way?

**Comment by ruslo** [comment]
> can I make the bgfx package default to these custom versions? and where do I define these versions?

@spacepluk We don't know a priori will user use bgfx or not (i.e. will hunter_add_package(bgfx) be called) so it's not possible to determine should we use "global default version" or "bgfx default version". For now this info available in documentation page and user have to decide is it ok to use custom versions or not. E.g. if you have some Lua stuff and you want to use Urho3D you have to lock yourself to Lua 5.1.5 or teach Urho3D to work with Lua 5.3.2 (which is default).

>  Since bgfx doesn't have tagged releases I was thinking of using the git SHAs of the packaged versions in the hunter_add_version calls. Is there a better way?

Anything stable should fit, e.g. you can take random commit from Hunter history:
* https://github.com/ruslo/hunter/archive/7c804a9241fbb31a3450caa0fecd1f68c18be3c4.zip

**Comment by spacepluk** [comment]
I've been taking a closer look at this, and packaging all the dependencies is going to be a lot of work and I'm not sure it's even worth it for most of them.

There are quite a few deps that are just fragments from other libraries or blog posts and I'm having a hard time finding all the original sources. For other cases where I did find the original, the current versions are so different that bgfx won't compile.

I suggest that we package bgfx, bx, and bimg without separating the dependencies unless they are already packaged in hunter and are compatible.  Then we can work from there and split the rest of the dependencies when there's an actual need to do it.  Otherwise I don't think it's going to be practical for anybody to package/maintain this.

What do you think?

**Comment by ruslo** [comment]
> I suggest that we package bgfx, bx, and bimg without separating the dependencies unless they are already packaged in hunter and are compatible.

That may work but it's not very robust. On both symbols layer and headers layer bundled dependencies may have conflicts with normal Hunter dependencies if user decide to mix them. I've wrote some details about how I see this problem in this comment: https://github.com/urho3d/Urho3D/pull/2170#issuecomment-352369375

**Comment by spacepluk** [comment]
Yeah, I understand that it's not ideal.  But in practice, that is only a problem if the dependency already exists in hunter, isn't it?

In those cases I can add the bgfx patched version to the existing package so that bgfx can use it (and work).  But I don't think it makes sense to package all the rest if nobody is going to use them, specially if they don't work with bgfx which is what I actually want to package.

That's why I'm saying that in my opinion it's better to have a bgfx package with some documented caveats, and then split the deps when there's a problem.

Regarding sending patches to upstream, I haven't asked but I doubt he's going to accept changing all the dependencies just because of CMake/hunter.

**Comment by ruslo** [comment]
> But in practice, that is only a problem if the dependency already exists in Hunter, isn't it?

If it's not in Hunter and if not same bundled approach used by another package.

> That's why I'm saying that in my opinion it's better to have a bgfx package with some documented caveats, and then split the deps when there's a problem.

Ok
